### PR TITLE
fix(ir): raise Exception for group by with no keys

### LINKED ISSUE #6237: bug: .distinct(on="does_not_exist") does not crash
### What happened?

```python
ibis.examples.topo.fetch().distinct(on="does_not_exist")
```

I would expect this to error.

### What version of ibis are you using?

5.1.0, have not tested on main

### What backend(s) are you using, if any?

duckdb, but probably irrelevant?

### Relevant log output

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by mesejo**
fixes #6237

### CODE REVIEW

**Comment by cpcloud** [comment]
Thanks for the PR @mesejo!

A `GROUP BY` with no keys is a valid expression in I think all of our SQL backends so I think we want to preserve that.

Would you mind addressing #6237 in a different way? How about fixing the issue directly in the `distinct` method? This would also allow us to craft a more specific error message that may be a bit more informative for end-users.

**Comment by cpcloud** [comment]
@mesejo Actually, I think I am mistaken!

I was thinking about something else and it's actually that `GROUP BY` is valid without any aggregates!

I think what you've done here is correct, so I'll review it as is.

File: ibis/expr/types/groupby.py [hunk 1]
```diff
@@ -79,6 +79,10 @@ def __init__(
                 ),
             )
         )
+
+        if not self.by:
+            raise ValueError("The grouping keys list is empty!")
```
Comment (cpcloud) [inline ibis/expr/types/groupby.py:84]: `ValueError` seems like it lacks useful type information. What about raising an `com.IbisInputError`, which while not super informative itself, it conveys that there was an error caused by some input from the user.

Comment (cpcloud) [inline ibis/expr/types/groupby.py:84]:
```suggestion
            raise ValueError("The grouping keys list is empty")
```

No need to shout :)

File: ibis/tests/expr/test_table.py [hunk 2]
```diff
@@ -1812,3 +1819,9 @@ def test_invalid_keep_distinct():
     t = ibis.table(dict(a="int", b="string"), name="t")
     with pytest.raises(com.IbisError, match="Invalid value for `keep`:"):
         t.distinct(on="a", keep="invalid")
+
+
+def test_invalid_distinct_empty_key():
+    t = ibis.table(dict(a="int", b="string"), name="t")
+    with pytest.raises(ValueError):
+        t.distinct(on="c", keep="first")
```
Comment (cpcloud) [inline ibis/tests/expr/test_table.py:1827]: I think we should still raise a more specific error message in `distinct(on="does_not_exist")`. Erroring in the group by that is the implementation of `distinct(on...)` leaks implementation details and may leave users wondering why a group by is involved.

Comment (cpcloud) [reply ibis/tests/expr/test_table.py:1827]: This can be done in a follow-up for sure!

**Review by cpcloud — CHANGES_REQUESTED** [review]

File: ibis/expr/types/groupby.py (hunk 1 above)
Comment (mesejo) [reply ibis/expr/types/groupby.py]: Agree, `com.IbisInputError` seems like a better option.

File: ibis/tests/expr/test_table.py (hunk 2 above)
Comment (mesejo) [reply ibis/tests/expr/test_table.py:1827]: I'll try to do it in a follow-up since I've to figure out how to do it.

**Comment by cpcloud** [comment]
The dask failure appears to be a flaky test related to ordering and the impala failure looks related to github actions.

**Comment by cpcloud** [comment]
Thanks!

**Pull request merged**